### ISSUE TITLE
Get conflict when deploy to API gateway in parallel via serverless?

### ISSUE THREAD

**Issue by zhaoyi0113** [serverless/serverless#9036]
I am deploying to AWS by serverless deploy command. There is a API Gateway already deployed and I deploy a couple of lambdas from different project to attach to the same API gateway.

The serverless.yml files for these projects look like:

provider:
  apiGateway:
    restApiId: !ImportValue ${self:custom.apiGatewayStackName}-apiGatewayId
    restApiRootResourceId: !ImportValue ${self:custom.apiGatewayStackName}- 
 apiGatewayRootResourceId
    metrics: true

function:
   ...
   events:
    - http:
        path: /customer
        integration: lambda
        method: post 
        request:
          template:
            application/json: '{"detail": $input.json("$")}'

so there are many different serverless.yml file which deploy the lambda to a different path and they are running together.

I got an random error during deploying these serverless.yml,  An error occurred: ApiGatewayDeployment1614293351030 - No integration defined for method (Service: AmazonApiGateway; Status Code: 400; Error Code: BadRequestException; Request ID: c90a3438-df7b-4c04-a3b6-d35eef1fcb9d; Proxy: null)..

This error happens randomly on different deploy script. They can be fixed by rerun the script. Also there is no issue if I run the script in sequence. So I believe the issue happens when we deploy these scripts in parallel. Is there any solution to fix it since running in sequence is very slow.

**Comment by pgrzesik** [comment]
Hello @zhaoyi0113 - thanks for reporting. Could you please follow the issue template that is presented to you when creating a new issue? It is helpful to have all details when trying to diagnose the problem. Did you experience at all this issue when running a single deployment? Do you have by any chance good steps to reproduce the problem?

**Comment by zhaoyi0113** [comment]
yes, I have deployed a API gateway and there are about 10 jobs who deploy some lambdas to this API gateway at the same time. Usually two or three of them failed with the error `ApiGatewayDeployment1614293351030 - No integration defined for method (Service: AmazonApiGateway; Status Code: 400; Error Code: BadRequestException; Request ID: c90a3438-df7b-4c04-a3b6-d35eef1fcb9d; Proxy: null)..`

they can be fixed by restarting the build job. And they are working perfect if I run these jobs in sequence. 
And also I have tried to use `cloudformation` instead of `serverless` to deploy them and they are working fine when they are running in parallel. Based on that I believe the issue is with `serverless` framework that why I created this bug in this repo.

**Comment by pgrzesik** [comment]
Thanks for clarification @zhaoyi0113 :bow: I'd love to dive into that problem but to be honest I don't think I will be able to in the coming days as there are some other priorities a bit higher in the queue and the issue itself seems like an edge case. I'll let you know as soon as I have a chance to dive deeper into it. Sorry about that.

**Comment by severi** [comment]
Since (most likely) today retrying the job has not worked for us anymore. When we retry the deployment with serverless it succeeds according to serverless, but might stay in a failed state on the AWS side

**Comment by pgrzesik** [comment]
Hello @severi - could you please share some more details about your specific case?

**Comment by severi** [comment]
We currently have ~13 sls stacks using the same api gateway (in a deployment step before these).

If all of the stacks get an update the deployment will basically always fail for around for of them. Looking at the cloudformation status this is what can be seen there

```
... UPDATE_ROLLBACK_COMPLETE

... UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS

my-sls-stack-name UPDATE_ROLLBACK_IN_PROGRESS

ApiGatewayDeployment12312312312	CREATE_FAILED

Resource handler returned message: "No integration defined for method (Service: ApiGateway, Status Code: 400, Request ID: xxxxxx, Extended Request ID: null)" (RequestToken: yyyy, HandlerErrorCode: InvalidRequest)
```

This has been a known issue that we were able to circumvent by just rerunning the failed deployments (Note: deployment does not build the sls functions as we build them in a previous step using `sls package`.

Yesterday on two different occasions and AWS accounts (read: environments) we realised that the redeploy did not actually rerun the deployment  event though serverless marked them as successfull - the status of the cloudformation for the failed stages was still `UPDATE_ROLLBACK_COMPLETE` and I doubt any deployment actions were actually ran there.

I eventually managed to get the problem fixed by rerunning the whole pipeline (including the actual building of the serverless stacks).

So in short - my gut feeling without actually knowing the inner working logic of serverless is that sls thinks that there's nothing to deploy and does not actually deploy the stacks  and then marks them as successfull.

This is somewhat weird as yesterday we had complete lambda functions / Api Gateway paths missing.

One more thing I realized was that when I rerun a deployment for a third time  for a stack that was missing lambdas - the stack actually got updated and fixed. Could it be that the second redeployment was run too fast ( I was kind of expecting problems so rerun it immediately when it failed) and there for the cloudformation stack was still in a "roll back" state and for some reason serverless thought the deployment succeeded. And then on the third time the roll back had finished and serverless was actually able to run the deployment correctly.

For stacks that did not introduce any new functions a simple redeploy did not help - we had to run the whole pipeline again including the build -step, which eventually fixed the these.

**Comment by pgrzesik** [comment]
> Yesterday on two different occasions and AWS accounts (read: environments) we realised that the redeploy did not actually rerun the deployment event though serverless marked them as successfull - the status of the cloudformation for the failed stages was still UPDATE_ROLLBACK_COMPLETE and I doubt any deployment actions were actually ran there.

Could you please include outputs of the `sls deploy` commands that were either failing or returned succesfully but didn't apply needed changes? What do you mean when you say `serverless marked them as successfull`?

> One more thing I realized was that when I rerun a deployment for a third time for a stack that was missing lambdas - the stack actually got updated and fixed. Could it be that the second (re)deployment was run too quickly ( I was kind of expecting problems so rerun it immediately when it failed) and at that point the cloudformation stack was still in a "roll back" state and for some reason serverless thought the deployment succeeded. And then on the third time the roll back had finished and serverless was actually able to run the deployment correctly.

If you try to update the stack if it's still updating or rolling back you should see an error.

**Comment by severi** [comment]
> > Yesterday on two different occasions and AWS accounts (read: environments) we realised that the redeploy did not actually rerun the deployment event though serverless marked them as successfull - the status of the cloudformation for the failed stages was still UPDATE_ROLLBACK_COMPLETE and I doubt any deployment actions were actually ran there.
> 
> Could you please include outputs of the `sls deploy` commands that were either failing or returned succesfully but didn't apply needed changes? What do you mean when you say `serverless marked them as successfull`?
> 

Here's an example in which the deployment fails (when multiple sls deploys are run at the same time)

```
$ yarn pnpify sls deploy --package .serverless.$CI_COMMIT_REF_NAME --stage $CI_COMMIT_REF_NAME
Serverless: Configuration warning at 'provider.tracing': should be object
Serverless:  
Serverless: Learn more about configuration validation here: http://slss.io/configuration-validation
Serverless:  
Serverless: Ensuring that deployment bucket exists
Serverless: Uploading CloudFormation file to S3...
Serverless: Uploading artifacts...
Serverless: Uploading service my-stack-name.zip file to S3 (5.57 MB)...
Serverless: Validating template...
Serverless: Updating Stack...
Serverless: Checking Stack update progress...
..................................................
Serverless: Operation failed!
Serverless: View the full error output: <url>
 
 Serverless Error ----------------------------------------
 
  An error occurred: ApiGatewayDeployment111111 - Resource handler returned message: "No integration defined for method (Service: ApiGateway, Status Code: 400, Request ID: xxxx, Extended Request ID: null)" (RequestToken: yyy, HandlerErrorCode: InvalidRequest).
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
 
  Your Environment Information ---------------------------
     Operating System:          linux
     Node Version:              14.15.5
     Framework Version:         2.70.0 (local)
     Plugin Version:            5.5.3
     SDK Version:               4.3.0
     Components Version:        3.18.1
 
Serverless: Deprecation warning: Syntax for referencing SSM parameters was upgraded with automatic type detection and there's no need to add "~true" or "~split" postfixes to variable references.
            Drop those postfixes and set "variablesResolutionMode: 20210326" in your service config to adapt to a new behavior.
            Starting with next major release, this will be communicated with a thrown error.
            
            More Info: https://www.serverless.com/framework/docs/deprecations/#NEW_VARIABLES_RESOLVER
Cleaning up project directory and file based variables
00:01
ERROR: Job failed: exit code 1
```

> > For stacks that did not introduce any new functions a simple redeploy did not help - we had to run the whole pipeline again including the build -step (sls package), which eventually fixed the these.
> 
> What was the command output for such redeploy that didn't fix the issue?

Don't have the output available anymore but luckily the problem happened again. Basically the interesting part is this:

`Serverless: Service files not changed. Skipping deployment...`

So serverless does not try to redeploy the stack eventhough on the AWS side the stack was marked as "UPDATE_ROLLBACK_COMPLETE" for the previous deployment

I assume this is the reason why the "second" delployment worked for the stack that was actually missing some of the lambdas (serverless recognized that the deployed lambdas did not match the current state and therefore redeployed the stack). As for the other stack that already had the same lambdas deployed (but with an older version) the actual deployment was skipped as "Service files have not changed".  Also if I had for example added other AWS resources using the `resources:` functionality they would not have gotten an update

**Comment by pgrzesik** [comment]
Thanks a lot for reporting further @severi - yes, this is a known current issue in the Framework, where the uploaded artifacts are not cleaned in case of rollbacks and that might cause these skipped deployments. We already track it as a part of bigger initiative here: https://github.com/serverless/serverless/issues/10436 but there are not immediate plans to take care of it so I cannot provide any ETA on the fix. Meanwhile, did you try to use `--force` flag to enforce deployment in such situations?

**Comment by severi** [comment]
We actually ended up splitting our single api gateway into multiple ones as described as Method 1 in this article https://seed.run/blog/how-to-share-an-api-domain-between-services-in-a-monorepo-serverless-app.html

I had previously been unaware of the possibility for multiple API Gateways to share the same base url using `basePath` options. With that basically all the issues seem to be fixed for us.

As a general note having a single Api Gateway for a monorepo project seems to be a bad idea (even though officially shouldn't cause any issues) as soon as the the gateway gets "big". Besides the issues mentioned here, AWS currently has an open issue in which you'll eventually encounter a "Template error: RootResourceId attribute of API Gateway RestAPI xxxxx doesn't exist" if you're API gateway has more than 200 resources attached to it.

The reason for the problem is the following (copy pasted from a response gotten from AWS)

```
... CloudFormation is making an API call with action “apiateway:GetResources” in order to get the root resource ID. This corresponds to the following API call [1]. This is a paginated response, meaning that it only returns resources 25 at a time (by default). Furthermore, to prevent throttling and time out issues, this is only checked a maximum of 8 pages deep. This means, that this root resource ID must be in the first 200 responses in order to get the root resource ID. Otherwise, it will fail to get this value.

...

The service team is aware of this issue and are working on a way to fix this. However, there is no ETA for this to be available. 
```

So my recommendation towards the issue would be to start using multiple api gateways if possible.

**Comment by pgrzesik** [comment]
Thanks for sharing that @severi 🙇

**Comment by mnapoli** [comment]
@severi on a different topic, we are exploring ideas for multi-service deployments (in monorepositories). We are looking for feedback on those ideas, it seems it would match your project here? I can share more details if you are interested, feel free to get in touch: by mail (matthieu.napoli at serverless dot com), or send me a message on the community slack ([link to join](https://join.slack.com/t/serverless-contrib/shared_invite/zt-13unxiht3-wTSFk910WilkYNpECmlPDA)).

**Comment by severi** [comment]
> @severi on a different topic, we are exploring ideas for multi-service deployments (in monorepositories). We are looking for feedback on those ideas, it seems it would match your project here? I can share more details if you are interested, feel free to get in touch: by mail (matthieu.napoli at serverless dot com), or send me a message on the community slack ([link to join](https://join.slack.com/t/serverless-contrib/shared_invite/zt-13unxiht3-wTSFk910WilkYNpECmlPDA)).

sure! I'll ping you on slack  🙂

**Comment by kim00425** [comment]
@severi thank you very much for your answer. currently I have the problem of ```Template error: RootResourceId attribute of API Gateway RestAPI xxxxx doesn't exist``` . So i'm going to do Method 1 like you. but now I have 300 functions. so should I split my project into all git repo? Is there any other way?

**Comment by ifeltsweet** [comment]
We've encountered a similar issue today however with a different error message:

```
Resource handler returned message: "Another Deployment is in progress for rest api with id xxxx . Please try again later. (Service: ApiGateway, Status Code: 409, Request ID: xxxx, Extended Request ID: null)" (RequestToken: xxxx, HandlerErrorCode: AlreadyExists)
```

We are deploying 11 services in parallel all targeting the same API Gateway. Deployments are done with Serverless Framework v3.17.0 using `deploymentMethod: direct`.

Folks from seed.run seem to have ran into this as well: https://seed.run/blog/auto-recovering-from-cloudformation-errors.html.
